internal/batcher/batcher.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2018 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package batcher supports batching of items. Create a Batcher with a handler and
+// add items to it. Items are accumulated while handler calls are in progress; when
+// the handler returns, it will be called again with the accumulated items.
+package batcher
+
+import (
+	"context"
+	"reflect"
+	"sync"
+)
+
+// A Batcher batches items.
+type Batcher struct {
+	handler       func(interface{}) error
+	itemSliceZero reflect.Value // nil (zero value) for slice of items
+	sem           chan struct{} // semaphore for outstanding handlers
+
+	mu      sync.Mutex
+	pending []waiter // items waiting to be handled
+}
+
+type waiter struct {
+	item interface{}
+	errc chan error
+}
+
+// New creates a new Batcher.
+//
+// itemExample is a value of the type that will be batched. For example, if you
+// want to create batches of *Entry, pass &Entry{} for itemExample.
+//
+// maxOutstandingHandlers is the maximum number of handlers that will run concurrently.
+//
+// handler is a function that will be called on each bundle. If itemExample is
+// of type T, the argument to handler is of type []T.
+func New(itemExample interface{}, maxOutstandingHandlers int, handler func(interface{}) error) *Batcher {
+	b := &Batcher{
+		handler:       handler,
+		sem:           make(chan struct{}, maxOutstandingHandlers),
+		itemSliceZero: reflect.Zero(reflect.SliceOf(reflect.TypeOf(itemExample))),
+	}
+	for i := 0; i < maxOutstandingHandlers; i++ {
+		b.sem <- struct{}{}
+	}
+	return b
+}
+
+// Add adds an item to the batcher. It blocks until the handler has
+// processed the item and reports the error that the handler returned.
+func (b *Batcher) Add(ctx context.Context, item interface{}) error {
+	// Create a channel to receive the error from the handler. Give it a buffer so
+	// that if a goroutine is in the middle of calling a handler when another
+	// finishes, the latter won't block sending to the channel.
+	c := make(chan error, 1)
+	b.mu.Lock()
+	b.pending = append(b.pending, waiter{item, c})
```

Comment:
I am concerned that this can grow unbounded, especially since there's no explicit deadline set on the batch sends. Implementing cancel should mitigate this, but at the very least, let's add a TODO.

Reply:
It can, but not in an interesting way (I claim).

Since `Add` is synchronous, in order to have N items in `pending` you need N goroutines. So you're already using memory proportional to the number of items. 

Put another way, you should be doing flow control with whatever is starting all those goroutines; the size of `pending` is just a side effect.